backend/capellacollab/projects/toolmodels/backups/runs/crud.py

```diff
@@ -52,7 +53,23 @@ def get_scheduled_and_running_pipelines(
     )
 
 
-def create_pipeline_run(db: Session, pipeline_run: models.DatabasePipelineRun):
+def create_pipeline_run(
+    db: orm.Session, pipeline_run: models.DatabasePipelineRun
+):
     db.add(pipeline_run)
     db.commit()
     return pipeline_run
+
+
+def get_last_pipeline_run_of_pipeline(
+    db: orm.Session, pipeline: pipeline_models.DatabaseBackup
+) -> models.DatabasePipelineRun | None:
+    return (
+        db.execute(
+            sqlalchemy.select(models.DatabasePipelineRun)
+            .where(models.DatabasePipelineRun.pipeline == pipeline)
+            .order_by(models.DatabasePipelineRun.id.desc())
+        )
+        .scalars()
+        .first()
+    )
```

Comment:
Personally, I would still prefer to just use the id. I'm not sure why we would need more than one foreign key between two objects. Also, using the whole object seems less flexible, since we always need the whole object from the database. So if we only have an id, we have to get the object from the database first before we can use this function (because of our dependency injection, the last point is not so bad for us).

Reply:
> Personally, I would still prefer to just use the id. I'm not sure why we would need more than one foreign key between two objects. Also, using the whole object seems less flexible, since we always need the whole object from the database. So if we only have an id, we have to get the object from the database first before we can use this function (because of our dependency injection, the last point is not so bad for us).

But it also prevents to just use the `pipeline_id` without further checks. For example, before calling crud functions, it has to be checked if the pipeline belongs to the tool-model and all permissions are alright (not for this specific case, because it's restricted to administrators, but in general). I don't see many use cases where we just have the pipeline id, but not the pipeline object. 

Another advantage of passing the pipeline object is typing. When just using `int` as type annotation, you can pass any id, for example also a model id, leading to wrong behaviour. When passing database objects, it should complain about mismatching types. 
(This is what happened with our read-only sessions where the `tool_id` was compared with the `version_id`, because it was passed as wrong argument in a call)

Multiple foreign keys are needed if you have multiple private keys on the references table. For example, this can be useful if you want to have pipeline specific IDs. In our current implementation, the pipeline run ID is counted globally, but when adding the pipeline ID as private key, it would only count up when a pipeline run for a specific pipeline is triggered. The pair to uniquely reference would be `(pipeline_id, pipeline_run_in)` then.

Reply:
> But one more point. Why do we need this function at all? If we have the `DatabaseBackup` object, can't we use `runs` directly and, if we need to be sure that `runs` is up to date, call `db.refresh(pipeline)` first?

I found no way to pass the `order_by` and `first` arguments when accessing the runs directly. Especially `first` is quite important as I don't want to fetch all runs if I just need the first one (this would be much more inefficient and may lead to a massive memory spike when the number of pipeline runs grows, because all runs would be loaded into memory)

Reply:
It would be good to align on this topic to also apply the decision consistently to all crud functions.